driver/src/contracts/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,255 @@
+            index,
+            account_state: AccountState::default(),
+            orders: Vec::new(),
+            user_order_counts: HashMap::new(),
+            pagination: Pagination {
+                previous_page_user: H160::zero(),
+                previous_page_user_offset: 0,
+            },
+        }
+    }
+
+    /// Applies one batch of data.
+    ///
+    /// A batch can come from `getEncodedUsersPaginated` or `getEncodedOrders`.
+    /// In the latter case there is only one batch.
+    ///
+    /// Returns the number of orders in the data.
+    ///
+    /// Panics if `packed_auction_bytes` is not a multiple of
+    /// `AUCTION_ELEMENT_WIDTH`.
+    pub fn apply_batch(&mut self, packed_auction_bytes: &[u8]) -> u64 {
+        let previous_order_count = self.orders.len();
+        self.apply_auction_data(&packed_auction_bytes);
+        let number_of_added_orders = (self.orders.len() - previous_order_count) as u64;
+        if number_of_added_orders == 0 {
+            return 0;
+        }
+        let last_order_user = self
+            .orders
+            .last()
+            .expect("have at least one order")
+            .account_id;
+        if self.pagination.previous_page_user == last_order_user {
+            self.pagination.previous_page_user_offset += number_of_added_orders;
+        } else {
+            self.pagination.previous_page_user = last_order_user;
+            self.pagination.previous_page_user_offset = number_of_added_orders;
+        }
+        number_of_added_orders
+    }
+
+    fn apply_auction_data(&mut self, packed_auction_bytes: &[u8]) {
+        let auction_elements = self.parse_auction_elements(packed_auction_bytes);
+        self.apply_auction_elements_to_account_state(auction_elements.iter());
+        self.orders.extend(
+            auction_elements
+                .into_iter()
+                .map(|auction_element| auction_element.order),
+        );
+    }
+
+    fn apply_auction_elements_to_account_state<'a, Iter>(&mut self, auction_elements: Iter)
+    where
+        Iter: Iterator<Item = &'a StableXAuctionElement>,
+    {
+        for element in auction_elements.into_iter() {
+            self.account_state.modify_balance(
+                element.order.account_id,
+                element.order.sell_token,
+                |x| *x = element.sell_token_balance,
+            );
+        }
+    }
+
+    fn parse_auction_elements(
+        &mut self,
+        packed_auction_bytes: &[u8],
+    ) -> Vec<StableXAuctionElement> {
+        assert_eq!(
+            packed_auction_bytes.len() % AUCTION_ELEMENT_WIDTH,
+            0,
+            "Each auction should be packed in {} bytes",
+            AUCTION_ELEMENT_WIDTH
+        );
+
+        // Workaround for borrow checker that would complain that the map
+        // closure borrows self as mutable while at the same time the filter
+        // closure borrows self as immutable because of using `self.index`.
+        let index = self.index;
```

Comment:
Could you capture by move in the filter condition instead?

Reply:
Making the filter closure `move` does not work if that is what you mean.

Reply:
Right, because we would move self and not just the index.

Reply:
You could user `filter_map` which would only mutably borrow `self` once.

Reply:
filter_map doesn't work because it applies the filter before the map but we actually perform a side effect in the filter updating `user_order_counts`.

Reply:
This would work but I find filter and map clearer.